test/e2e/jmx/jmx_test.go

```diff
@@ -0,0 +1,364 @@
+		if instance.InstanceName != nil {
+			nodeNames = append(nodeNames, *instance.InstanceName)
+		}
+	}
+
+	os.Exit(m.Run())
+}
+
+//------------------------------------------------------------------------------
+// Main Test Functions
+//------------------------------------------------------------------------------
+
+func TestAll(t *testing.T) {
+	t.Run("Resources", func(t *testing.T) {
+		testResources(t)
+	})
+
+	// Don't run metric tests if resource tests fail
+	if !t.Failed() {
+		t.Run("Metrics", func(t *testing.T) {
+			testMetrics(t)
+		})
+	}
+}
+
+func testResources(t *testing.T) {
+	tests := testResourcesRegistry
+
+	config, err := clientcmd.BuildConfigFromFlags("", filepath.Join(os.Getenv("HOME"), ".kube", "config"))
+	require.NoError(t, err, "Error building kubeconfig")
+
+	clientset, err := kubernetes.NewForConfig(config)
+	require.NoError(t, err, "Error creating clientset")
+
+	for _, testFunc := range tests {
+		testFunc(t, clientset)
+	}
+}
+
+func testMetrics(t *testing.T) {
+	configFile := filepath.Base(env.AgentConfig)
+	tests := testMetricsRegistry[configFile]
+
+	fmt.Println("waiting for metrics to propagate...")
+	time.Sleep(wait)
+
+	for _, testFunc := range tests {
+		testFunc(t)
+	}
+}
+
+//------------------------------------------------------------------------------
+// Resource Test Functions
+//------------------------------------------------------------------------------
+
+func testAgentResources(t *testing.T, clientset *kubernetes.Clientset) {
+	t.Run("verify_agent_resources", func(t *testing.T) {
+		daemonSet, err := clientset.AppsV1().DaemonSets("amazon-cloudwatch").Get(context.TODO(), "cloudwatch-agent", metav1.GetOptions{})
+		require.NoError(t, err, "Error getting CloudWatch Agent DaemonSet")
+		require.NotNil(t, daemonSet, "CloudWatch Agent DaemonSet not found")
+
+		configMap, err := clientset.CoreV1().ConfigMaps("amazon-cloudwatch").Get(context.TODO(), "cloudwatch-agent", metav1.GetOptions{})
+		require.NoError(t, err, "Error getting CloudWatch Agent ConfigMap")
+		require.NotNil(t, configMap, "CloudWatch Agent ConfigMap not found")
+
+		cwConfig, exists := configMap.Data["cwagentconfig.json"]
+		require.True(t, exists, "cwagentconfig.json not found in ConfigMap")
+		require.Contains(t, cwConfig, `jmx`, "JMX configuration not found in cwagentconfig.json")
+
+		service, err := clientset.CoreV1().Services("amazon-cloudwatch").Get(context.TODO(), "cloudwatch-agent", metav1.GetOptions{})
+		require.NoError(t, err, "Error getting CloudWatch Agent Service")
+		require.NotNil(t, service, "CloudWatch Agent Service not found")
+
+		serviceAccount, err := clientset.CoreV1().ServiceAccounts("amazon-cloudwatch").Get(context.TODO(), "cloudwatch-agent", metav1.GetOptions{})
+		require.NoError(t, err, "Error getting CloudWatch Agent Service Account")
+		require.NotNil(t, serviceAccount, "CloudWatch Agent Service Account not found")
+	})
+}
+
+func testJMXResources(t *testing.T, clientset *kubernetes.Clientset) {
```

Comment:
Can this functionality be made generic ? Feels like its checking pods status, env vars which could be re-used for other functionality testing as well

Reply:
Yeah, I agree that it could be made more generic. I can adjust for that.